pkg/build/api/v1/types.go

```diff
@@ -163,6 +163,9 @@ type BuildSource struct {
 	// repository.
 	ContextDir string `json:"contextDir,omitempty" description:"specifies sub-directory where the source code for the application exists, allows for sources to be built from a directory other than the root of a repository"`
 
+	// DockerfileName is the name of the Dockerfile that will be used to build the Docker image.
+	DockerfileName string `json:"dockerfileName,omitempty" description:"specifies the name of the Dockerfile to use for building the Docker image"`
```

Comment:
> Totally agree, but we already do that with BuildSource and *Strategy structs.

example?  I don't see anything i'd consider to be an "input" in the strategy structs, so this would be the first.  the only thing you could sort of argue is an input is the pull secret + builder image.


> Unfortunately since this is done in fetchSource method used with docker and sti builder you may end up with unexpected behavior, where the Dockerfile will be overwritten in source strategy,

no, you won't, because validation prevents that, which is what I pointed out in my original comments on this..that we should be using a single generic object with validation.. "if we're not going to do that, i think we should continue down the path we've established of having a single generic buildsource object which contains the superset of inputs that our strategies can understand, and handle invalid values via validation.":
https://github.com/openshift/origin/blob/master/pkg/build/api/validation/validation.go#L122-L126

So again, given where we're at, I prefer we keep a single generic "input" object, which is shared by all strategies, and validation logic which ensures the input content makes sense for the chosen strategy.

at some point when we can afford to refactor the api, i'd be ok with having an input-type struct for each specific build strategy.

but i do not want to start putting input content into build strategies today, and i consider the dockerfilename to be input content.

After all, what if someone using a custom builder would also like to use a dockerfilename?  sure, they can use an env variable to provide, it, but why shouldn't they be able to provide it in a first class way?  so now you'd have to introduce a dockerfilename field in the custom build strategy struct too?